### PR TITLE
Check mysql charset, if it's utf8, display an error log

### PR DESCRIPTION
**Pull request by lunny**
This PR will check MySQL database charset when Gitea start. If it's `utf8`, an error log will be displayed.

### CODE REVIEW

**Comment by wolfogre** [comment]
The CI failure is related.

image

Hmm, so we used an incorrect database to test for years? 😂

**Comment by lunny** [comment]
It's wired, we never test emoji with our CI mysql services?

**Comment by wxiaoguang** [comment]
What about adding a "self check" page in site admin?

It could do a full self check, including:

1. database versions (pgsql, mysql, etc)
2. database charsets (mysql)
3. reverse proxy rewriting and sub path (eg: is `%2F` kept?)
4. system time (to resolve incorrect OTP problems)
5. system health like free disk, etc

**Comment by lunny** [comment]
> What about adding a "self check" page in site admin?
> 
> It could do a full self check, including:
> 
>     1. database versions (pgsql, mysql, etc)
> 
>     2. database charsets (mysql), not only just checking default, but check all columns.
> 
>     3. reverse proxy rewriting and sub path (eg: is `%2F` kept?)
> 
>     4. system time (to resolve incorrect OTP problems)
> 
>     5. system health like free disk, etc

sounds like web-based gitea doctor

**Comment by wxiaoguang** [comment]
Yes, some checks could only be done by web UI (eg: 3, 4)

**Comment by silverwind** [comment]
UI self check I'd say is out of scope for this PR.

File: models/db/engine.go [hunk 1]
```diff
@@ -147,6 +148,25 @@ func InitEngine(ctx context.Context) error {
...
+		setting.Database.Name).Get(&dbCharset)
+	if err != nil {
+		return err
+	}
+	if dbCharset != setting.Database.Charset {
+		return fmt.Errorf("database setting charset is %s but in fact it's %s", setting.Database.Charset, dbCharset)
+	}
+	if dbCharset == "utf8" {
+		log.Error("database charset is utf8 which is deprecated, please convert it to utf8mb4 and change your app.ini [database]charset")
+	}
```
Comment (silverwind) [inline models/db/engine.go:166]: Also match for `utf8mb3` which is an alias to `utf8` and interpolate charset into the message.

Reference: https://dev.mysql.com/doc/refman/8.0/en/charset-unicode-utf8.html

**Comment by wxiaoguang** [comment]
> UI self check I'd say is out of scope for this PR, but I agree it might be useful.

But checking default charset is not enough. Every column should be checked one by one.

**Comment by lunny** [comment]
> > UI self check I'd say is out of scope for this PR, but I agree it might be useful.
> 
> But checking default charset is not enough. Every column should be checked one by one.

I think you are right, but we should not check them when start. It should be a backend task.

**Comment by wxiaoguang** [comment]
> > > UI self check I'd say is out of scope for this PR, but I agree it might be useful.
> > 
> > 
> > But checking default charset is not enough. Every column should be checked one by one.
> 
> I think you are right, but we should not check them when start. It should be a backend task.

That why I suggest to make an admin UI

**Comment by silverwind** [comment]
Still not sure about UI.

How about actually refusing to start on wrong charset? Yes it's strict, but it's the only way to really ensure all users get it.

**Comment by wxiaoguang** [comment]
IMO an admin UI is the only solution. See these 5 points (actually not only these)

We shouldn't stop Gitea from starting, and we should do full check (check every column) when starting. So we still need an admin UI.

**Comment by silverwind** [comment]
How about doing a full table check and display a warning on dashboard to every admin user if issues found? That should be a simple UI to add and would be better than something hidden away in the admin section. I assume a full table scan won't take so long.

**Comment by techknowlogick** [comment]
> How about doing a full table check and display a warning on frontpage to every admin user if issues were found? That should be a simple UI to add and would be better than something hidden away in the admin section. I assume a full table scan won't take so long.
> 
> For people concerned about startup performance, we could add a opt-out to these checks in config.

Atlassian does exactly this (a full table check, and then reports in UI) for the datacentre version of their products, they aren't blocking on the boot of the server (IIRC) and so we could probably do something similar.

I agree re: having it block the startup for the check is bad. At the very least a one-off "on-boot" cron might be sufficient.

**Comment by lunny** [comment]
Close as this will break users' instances.

**Pull request closed without merging**